okhttp/src/main/java/io/grpc/okhttp/OkHttpFrameLogger.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.okhttp;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.okhttp.internal.framed.ErrorCode;
+import io.grpc.okhttp.internal.framed.Header;
+import io.grpc.okhttp.internal.framed.Settings;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import okio.Buffer;
+import okio.ByteString;
+
+class OkHttpFrameLogger {
+  private static final int BUFFER_LENGTH_THRESHOLD = 64;
+  private final Logger logger;
+  private final Level level;
+
+  OkHttpFrameLogger(Level level, Class<?> clazz) {
+    this(level, Logger.getLogger(clazz.getName()));
+  }
+
+  private OkHttpFrameLogger(Level level, Logger logger) {
+    this.level = checkNotNull(level, "level");
+    this.logger = checkNotNull(logger, "logger");
+  }
+
+  private static String toString(Settings settings) {
+    EnumMap<SettingParams, Integer> map = new EnumMap<>(SettingParams.class);
+    for (SettingParams p : SettingParams.values()) {
+      // Only log set parameters.
+      if (settings.isSet(p.getBit())) {
+        map.put(p, settings.get(p.getBit()));
+      }
+    }
+    return map.toString();
+  }
+
+  private String toString(ByteString buf) {
+    if (level.intValue() <= Level.FINE.intValue() || buf.size() <= BUFFER_LENGTH_THRESHOLD) {
```

Comment:
I don't understand why you check level? Isn't it always `Level.FINE`? Why do you log entire buffer in that case? Is it the same behavior as Netty?

Reply:
Sure. I will make the change.